### DISCUSSION TITLE
What is the purpose of `HelmRepository.spec.interval` when there is a `HelmChart.spec.interval`

### DISCUSSION THREAD

**Discussion by VTimNeumann** [fluxcd/flux2#4721] [Q&A] [answered]
I was wondering what the purpose of the required `HelmRepository.spec.interval` when all Helm Charts also have an interval (by setting `HelmRelease.spec.chart.spec.interval` or leaving it empty, which causes it to default to `HelmRelease.spec.interval`).

As I understand, both are used to check for updates of the charts in the repository.
Obviously the interval on each chart governs the update checking for that chart only, whereas the one on the repository does check for updates in the whole repository. 
But as all charts have such an interval, it seems to me the one on the repository is redundant.

See https://fluxcd.io/flux/components/source/api/v1beta2/#source.toolkit.fluxcd.io/v1beta2.HelmRepository
See https://fluxcd.io/flux/components/helm/api/v2beta2/#helm.toolkit.fluxcd.io/v2beta2.HelmChartTemplate

**Comment by souleb** [comment, marked as answer]
The `interval` is used as a `reconciliation` period. Each custom resource is reconciled by a different controller. 
For a `HelmRepository` the principal action is retrieving and persisting the `helm index`.
For a `HelmChart` it is about retrieving the right `chart` based on the `helm index` provided by the corresponding `HelmRepository`. Different intervals will give different results as you may have a different view of the `helm repository`.

**Reply by VTimNeumann** [reply]
Thanks for the reply.
Does that mean that, given a new chart version is released, the reconciliation of the `HelmChart` does not update the chart to that newest available version unless the reconciliation of the `HelmRepository` ran between the release and the reconciliation of the `HelmChart`?

**Reply by souleb** [reply]
Yes, the `index` file list available versions, so it must be up to date. You can go around this by using a `HelmRepository` of type `oci`, where the controller will always request available versions(tags) from the registry api.

**Reply by VTimNeumann** [reply]
Ok thanks a lot. One more question: Assuming I'm not using an OCI repo, is there any point in setting the `HelmChart` reconciliation to be more frequent than the `HelmRepository` reconciliation? (I've seen tutorials where 12h for the chart and 24h for the repo have been used)

**Reply by souleb** [reply]
I'd make the `HelmRepository` frequency lower to make sure to catch updates before the chart reconciliation.

**Reply by VTimNeumann** [reply]
Do you mean you'd have a lower interval for the repo?
I.e. `HelmRepository.spec.interval: 6h` and `HelmRelease.spec.chart.spec.interval: 12h`
Or did you actually mean a lower frequency => higher period == higher interval for the repo?
I.e.  `HelmRepository.spec.interval: 24h` and `HelmRelease.spec.chart.spec.interval: 12h`

**Reply by souleb** [reply]
Oh sorry, yes lower interval.

**Reply by antonblr** [reply]
When:
* `HelmRepository` sync interval is 6h
* `HelmRelease` sync interval is 12h

Assuming I just build and pushed a new Helm release.

What would happen in case if I update and push a new `HelmRelease.spec.chart.spec.version` to the Git repository? 
From my understanding it will try to reconcile `HelmRelease` right away (because of change), but it won't find a new Helm release version until `HelmRepository` reconciliation loop happens, where it updates the index (which can be up to 6 hours). Is this correct assumption?

Or, perhaps `HelmRepository`  will be reconciled in the same wave, because `notification-contoller` will send an event of Git repo push?
